[강진호] week8

jinhok96 opened the pull request:
## 요구사항

### 기본

- [x] 링크 공유 페이지 url path는 ‘/shared’, 폴더 페이지 url path는 ‘/folder’가 되도록 설정했나요?
- [x] 폴더 페이지에서 상단 네비게이션 바는 스크롤 시 상단에 고정하지 않고 가려지도록 했나요?
- [x] 상단 네비게이션 바에 프로필 영역의 데이터는 https://bootcamp-api.codeit.kr/docs 에 명세된 “/api/users/1”을 활용했나요?
- [x] “전체” 폴더를 선택한 경우 “공유”, “이름 변경”, “삭제” 버튼들이 보이지 않지만, 다른 폴더를 선택한 경우에는 버튼들이 보이나요?
- [x] 폴더 목록에 필요한 데이터는 “/api/users/1/folders”를 활용했나요?
- [x] “전체” 폴더에 필요한 링크들 데이터는 “/api/users/1/links”를 활용하고, 이외의 폴더에 필요한 링크들 데이터는 “/api/users/1/links?folderId={해당 폴더 ID}”를 활용했나요?
- [x] Mobile에서 “폴더 추가” 버튼은 최하단에서 101px 떨어져있는 Floating Action Button으로 만들었나요?

## 주요 변경사항

- 모든 기능 구현 완료했습니다.
- 1차 코드리뷰 피드백 전부 반영했습니다.

## 스크린샷

![image](이미지url)

## 멘토에게

- 주소창에 /shared, /folder로 접근 가능합니다.
- 셀프 코드 리뷰를 통해 질문 이어가겠습니다.

jinhok96 committed:
new file:   .eslintrc.js
	modified:   .gitignore
	new file:   .prettierrc

jinhok96 committed:
chore: Shared 페이지 이름 SharedPage로 변경
	modified:   src/App.js
	renamed:    src/pages/Shared/Shared.css -> src/pages/SharedPage/SharedPage.css
	renamed:    src/pages/Shared/Shared.js -> src/pages/SharedPage/SharedPage.js

jinhok96 committed:
chore: HomePage, FolderPage 추가
	new file:   src/pages/FolderPage/FolderPage.css
	new file:   src/pages/FolderPage/FolderPage.js
	new file:   src/pages/HomePage/HomePage.css
	new file:   src/pages/HomePage/HomePage.js

jinhok96 committed:
chore: 페이지 주소 변경
	modified:   src/App.js

jinhok96 committed:
chore: FolderPage에 임시 콘텐츠 추가
	modified:   src/pages/FolderPage/FolderPage.js

jinhok96 committed:
fix: 잘못된 margin 수정
	modified:   src/components/Header/Header.module.css

jinhok96 committed:
chore: link 아이콘 추가
	new file:   src/assets/images/link.svg

jinhok96 committed:
chore: AddLinkBar 컴포넌트 추가
	new file:   src/components/Header/AddLinkBar.js
	new file:   src/components/Header/AddLinkBar.module.css

jinhok96 committed:
chore: AddLInkButton 추가
	modified:   src/assets/styles/colors.css
	new file:   src/components/Common/AddLInkButton.js
	new file:   src/components/Common/AddLInkButton.module.css
	modified:   src/components/Common/LoginButton.module.css
	modified:   src/components/Header/AddLinkBar.js
	modified:   src/components/Header/AddLinkBar.module.css

jinhok96 committed:
chore: FolderHeaderContent 컴포넌트 추가
	new file:   src/components/Header/FolderHeaderContent.js
	new file:   src/components/Header/FolderHeaderContent.module.css

jinhok96 committed:
feat: AddLinkBar 추가
	modified:   src/components/Header/AddLinkBar.js
	modified:   src/components/Header/AddLinkBar.module.css
	modified:   src/components/Header/FolderHeaderContent.js
	modified:   src/components/Header/FolderHeaderContent.module.css
	modified:   src/components/Header/SharedHeaderContent.module.css
	modified:   src/pages/FolderPage/FolderPage.js

jinhok96 committed:
fix: 리액트상 여러가자 오류 수정
	modified:   src/components/Footer/Footer.js
	modified:   src/pages/HomePage/HomePage.js

jinhok96 committed:
chore: AddLInkButton 추가 및 AddLinkBar 크기 조정
	modified:   src/components/Common/AddLInkButton.module.css
	modified:   src/components/Header/AddLinkBar.js
	modified:   src/components/Header/AddLinkBar.module.css

jinhok96 committed:
refactor: Main 마진 재분배
	modified:   src/components/Header/Header.module.css
	modified:   src/components/Main/CardList.module.css
	modified:   src/components/Main/Main.module.css
	modified:   src/components/Main/SearchBar.js
	modified:   src/components/Main/SearchBar.module.css

jinhok96 committed:
fix: 잘못된 이름 수정
	modified:   src/pages/FolderPage/FolderPage.js

jinhok96 committed:
chore: SortingButton 추가
	new file:   src/components/Common/SortingButton.js
	new file:   src/components/Common/SortingButton.module.css

jinhok96 committed:
feat: SortingButtonList 추가
	new file:   src/assets/images/add.svg
	new file:   src/assets/images/delete.svg
	new file:   src/assets/images/pen.svg
	new file:   src/assets/images/share.svg
	modified:   src/assets/styles/colors.css
	modified:   src/assets/styles/utility.css
	modified:   src/components/Common/Button.js
	new file:   src/components/Common/FlaotingActionButton.js
	new file:   src/components/Common/FloatingActionButton.module.css
	modified:   src/components/Common/SortingButton.js
	modified:   src/components/Common/SortingButton.module.css
	new file:   src/components/Main/SortingButtonList.js
	new file:   src/components/Main/SortingButtonList.module.css
	modified:   src/pages/FolderPage/FolderPage.js

jinhok96 committed:
chore: FAB에 className prop 추가
	modified:   src/components/Common/FlaotingActionButton.js

jinhok96 committed:
chore: AddFolderButton 추가
	new file:   src/components/Common/AddFolderButton.js
	new file:   src/components/Common/AddFolderButton.module.css

jinhok96 committed:
chore: AddFolderButton 스타일 추가
	modified:   src/assets/styles/colors.css
	modified:   src/components/Common/AddFolderButton.js
	modified:   src/components/Common/AddFolderButton.module.css

jinhok96 committed:
chore: AddFolderButton 스타일 수정
	modified:   src/assets/images/add.svg
	modified:   src/components/Common/AddFolderButton.js
	modified:   src/components/Common/AddFolderButton.module.css
	modified:   src/components/Common/Button.js
	modified:   src/pages/FolderPage/FolderPage.js

jinhok96 committed:
chore: AddFolderButton 화면에 출력
	modified:   src/components/Main/SortingButtonList.module.css
	modified:   src/pages/FolderPage/FolderPage.js
	new file:   src/pages/FolderPage/FolderPage.module.css
	modified:   src/pages/HomePage/HomePage.js
	renamed:    src/pages/FolderPage/FolderPage.css -> src/pages/HomePage/HomePage.module.css
	deleted:    src/pages/SharedPage/SharedPage.css
	modified:   src/pages/SharedPage/SharedPage.js
	renamed:    src/pages/HomePage/HomePage.css -> src/pages/SharedPage/SharedPage.module.css

jinhok96 committed:
chore: AddFolderButton 반응형 속성 추가
	modified:   src/pages/FolderPage/FolderPage.js
	modified:   src/pages/FolderPage/FolderPage.module.css

jinhok96 committed:
chore: 버튼 선택 시 스타일 변경 추가
	modified:   src/components/Main/SortingButtonList.js
	modified:   src/components/Main/SortingButtonList.module.css

jinhok96 committed:
chore: selectedButton 초기값 지정
	modified:   src/components/Main/SortingButtonList.js

jinhok96 committed:
feat: Sorting Section 스타일링 완료
	new file:   src/assets/images/add-gray10.svg
	renamed:    src/assets/images/add.svg -> src/assets/images/add-primary.svg
	modified:   src/assets/styles/colors.css
	modified:   src/components/Common/AddFolderButton.js
	deleted:    src/components/Common/FlaotingActionButton.js
	new file:   src/components/Common/FlaotingAddFolderButton.js
	renamed:    src/components/Common/FloatingActionButton.module.css -> src/components/Common/FloatingAddFolderButton.module.css
	new file:   src/components/Common/Option.js
	new file:   src/components/Common/Option.module.css
	modified:   src/components/Main/SortingButtonList.js
	modified:   src/components/Main/SortingButtonList.module.css
	modified:   src/pages/FolderPage/FolderPage.js
	modified:   src/pages/FolderPage/FolderPage.module.css

jinhok96 committed:
refactor: 페이지 이름 변경
	renamed:    src/components/Main/SortingButtonList.js -> src/components/Main/SortingSection.js
	renamed:    src/components/Main/SortingButtonList.module.css -> src/components/Main/SortingSection.module.css
	modified:   src/pages/FolderPage/FolderPage.js

jinhok96 committed:
chore: gnb가 /folder 페이지에서 fixed 속성이 제거되도록 변경
- fixed일 경우 더미 공간을 생성하도록 수정
	modified:   src/components/Header/Gnb.js
	modified:   src/components/Header/Gnb.module.css
	modified:   src/components/Header/Header.js
	modified:   src/components/Header/Header.module.css

jinhok96 committed:
chore: 커스텀 훅 추가
	new file:   src/hooks/useFetch.js

jinhok96 committed:
chore: api.js에 커스텀 훅 적용
	modified:   src/services/api.js

jinhok96 committed:
chore: gnb에 변경된 api 적용
	modified:   src/components/Header/Gnb.js
	modified:   src/services/api.js

jinhok96 committed:
feat: CardList에 api 1차 적용
	modified:   src/components/Header/SharedHeaderContent.js
	modified:   src/components/Main/CardList.js
	modified:   src/hooks/useFetch.js
	modified:   src/services/api.js

jinhok96 committed:
chore: api 주소 변경
	modified:   src/services/api.js

SeolJaeHyeok commented on .eslintrc.js:
```diff
@@ -0,0 +1,48 @@
+    'max-len': ['error', 120, 2, { ignoreComments: true, ignoreUrls: true }],
+    'no-console': 'off',
+  },
+};
```
👍👍👍

SeolJaeHyeok commented on src/App.js:
```diff
@@ -1,12 +1,16 @@
+        <Route path="/" element={<HomePage />} />
+        <Route path="shared" element={<SharedPage />} />
+        <Route path="folder" element={<FolderPage />} />
       </Routes>
```
route 컴포넌트를 를 개별적으로 선언하는 것도 좋지만 path와 element를 별도의 상수로 관리하고 <Routes> 컴포넌트 내부에서 map으로 렌더링 시켜주는게 관리 측면에서 더 좋습니다!

react-router-dom v6에 새로 나온 [createBrowserRouter](https://reactrouter.com/en/main/routers/create-browser-router)를 이용해봐도 좋을거 같아요:)

SeolJaeHyeok commented on src/components/Common/AddFolderButton.js:
```diff
@@ -0,0 +1,24 @@
+import AddIcon from '../../assets/images/add-primary.svg';
+
+function AddFolderButton({ className, onClick }) {
+  const buttonClasses = classNames(
+    styles['add-folder-button'],
+    'text-color-primary',
+    'background-white',
+    'flex-row',
+    className
+  );
+  const iconClasses = classNames(styles['add-folder-icon']);
```
`classnames` 모듈을 추가해서 사용하신 이유가 궁금해요~!

jinhok96 replied:
assets/styles에 분리해놓은 여러 공통 스타일링을 사용하고 있는데, 이 경우 적용하는 클래스가 많아지다보니 여러  클래스명을 길게 작성할 경우 가독성이 좋지 않은 문제가 있었습니다. 그래서 classnames 모듈을 추가하여 클래스명을 관리하게 되었고 전체적인 코드 스타일을 통일시키기 위해 모든 클래스명을 classnames로 작성하고 있습니다! classnames는 이전 코드리뷰에서 알려주신 [css module](https://www.tcpschool.com/react/react_styling_cssmodule) 문서에서 처음 확인하고 적용했습니다!

SeolJaeHyeok commented on src/components/Common/Button.js:
```diff
@@ -1,12 +1,18 @@
+function Button({ className, text, onClick, children }) {
+  const buttonClasses = classNames(styles.button, className);
+
+  const button = (
+    <button className={buttonClasses} type="button" onClick={onClick}>
       {text}
+      {children}
     </button>
   );
```
children 👍

SeolJaeHyeok commented on src/components/Header/AddLinkBar.js:
```diff
@@ -0,0 +1,37 @@
+
+function AddLinkBar() {
+  const [value, setValue] = useState('');
+
```
변수 이름은 최대한 구체적으로 작성해서 변수명만 가지고 어떤 값인지 유추할 수 있게 작성하는걸 추천드립니다!
변수 이름 짓는건 개발자들이 가장 중요하게 생각하는 요소 중에 하나라서 의미 있는 변수 명을 작성하는 습관을 들여주세용!

SeolJaeHyeok commented on src/components/Header/Gnb.js:
```diff
@@ -1,22 +1,26 @@
 import LoginButton from '../Common/LoginButton';
 import logo from '../../assets/images/logo.svg';
 import ErrorMessage from '../Common/ErrorMessage';
+import DefaultProfileImg from '../../assets/images/profile-img.png';
```
파일 구조가 복잡해지면 `import` 문에 상대 경로들이 계속해서 길어지는 현상이 발생하고 가독성을 해치게 됩니다.
이런 경우 발생할 수 있는 대표적인 문제가 (에디터가 자동으로 변경된 경로를 잡아주는 경우도 있지만) 파일 구조가 변경됐을 떄 일일이 경로를 변경해줘야 하는 상황이 생겨요.

만약 절대 경로로 `import`를 하게되면  추후 파일 위치가 이동해도 import를 다시 맞추지 않아도 되는 장점이 있습니다.
그리고 유의미할 지는 모르겠으나 성능 상의 이점도 있다고 하네요!

진호님은 ESLint, Prettier도 잘 설정해주셔서 절대 경로 설정까지 한 번 진행해보셨으면 좋겠어요!

절대 경로를 설정하려면 `jsconfig.json` 혹은 `tsconfig.json`을 수정해야 하는데 이 부분은 내용이 많아서 추후에 한 번 정리해서 이야기를 드리는걸로 하고 지금은 우선 `react 절대 경로`와 같은 키워드로 검색해서 적용해보시면 될 거 같아요~

SeolJaeHyeok commented on src/components/Header/Gnb.js:
```diff
@@ -1,22 +1,26 @@
 // 글로벌 네비게이션 바
 function Gnb() {
+  const FOLDERLOCATION = '/folder';
+
```
무조건적인 규칙은 아니지만 상수를 선언할 때는 일반적으로 UPPER_SNAKE_CASE로 작성해준답니다!

SeolJaeHyeok commented on src/components/Header/Gnb.js:
```diff
@@ -1,22 +1,26 @@
   const [userInfo, setUserInfo] = useState(null);
   const [errorMessage, setErrorMessage] = useState('');
+  const location = useLocation().pathname;
 
```
여기도 변수명이 조금 어색한거 같아요. `pathname` 값을 추출하는 용도였다면 변수명이 `pathname`이 되어야 할 것 같습니다. 

혹은 만약 나중에 location 안의 또다른 값을 사용하게 된다고 생각하면 좀 더 수정이 편리할 수 있게 아래와 같이 destructuring 하는게 좋아보여요!

```javascript
    const { pathname, state, search } = useLocation()
```

SeolJaeHyeok commented on src/components/Header/Gnb.js:
```diff
@@ -31,33 +35,39 @@ function Gnb() {
 
-  const userProfileImg = userInfo?.profileImageSource || '';
+  const userProfileImg = userInfo?.profileImageSource || DefaultProfileImg;
   const userEmail = userInfo?.email || '';
```
논리 연산자(`||`)를 사용하는 것도 좋지만 널 병합 연산자(`??`)를 사용하는게 더 안전한 방식이에요!
아래 사진을 보면 아시겠지만 논리연산자는 Falsy 값을 처리하는게 널 병합 연산자와 다릅니다.

![image](https://github.com/codeit-bootcamp-frontend/4-Weekly-Mission/assets/71430291/38d8413f-5dc3-4e45-b5a1-e9499f35ec1b)

논리 연산자는 Falsy값을 false로 처리하기 때문에 의도치 않은 결과가 발생할 수 있어요.
Falsy 값은 자바스크립트에서 헷갈리기 쉬운 개념이기도 하고 개발을 복잡하게 만드는 요소니까 안쓰는게 좋습니다.

SeolJaeHyeok commented on src/components/Header/SharedHeaderContent.js:
```diff
@@ -12,7 +12,7 @@ function SharedHeaderContent() {
   const handleLoadFolder = async () => {
     try {
-      const result = await getFolderInfo();
+      const result = await getFoldersInfo();
       setFolderInfo(result);
```
이건 사람마다 견해가 다를 수 있는데, 저는 개인적으로 배열 데이터는 `s`를 붙여서 복수형을 나타내지 않고 `~List`와 같이 표현을 해주고 있어요.

이유는 영어에서 복수형을 표현할 때 단순히 `s`를 붙이면 되는 경우와 `es`가 붙는 경우 처럼 여러 경우가 있어서 변수로 표현하기 애매한 경우가 생기더라구요 ㅎㅎ 

진호님은 어떤게 더 나아 보이실까요?!?

jinhok96 replied:
저도 List가 알아보기 좋아서 List로 사용합니다! Folders는 api 주소가 folders라서 그대로 사용했습니다!

SeolJaeHyeok commented on src/hooks/useFetch.js:
```diff
@@ -0,0 +1,19 @@
+  const response = await fetch(url);
+  const result = await response.json();
+  const responseError = result.error;
+
```
성공했을 때 `result` 객체에 `error`가 없을 수도 있는데, 이런 경우는 어떻게 처리해야 될까요?

SeolJaeHyeok commented on src/hooks/useFetch.js:
```diff
@@ -0,0 +1,19 @@
+const API_URL = 'https://bootcamp-api.codeit.kr/api/';
+
```
API URL을 변수로 분리하는건 좋았는데, 이왕이면 환경 변수로 분리해볼까요?

SeolJaeHyeok reviewed (commented):
8주차 미션도 고생많으셨습니다 진호님 커밋도 잘 나눠주셨고 컴포넌트도 잘 분리해주셨고 이번 미션은 정말 잘 만들어주신 것 같아요!👏👏👏
드리고 싶은 말은 코드 리뷰로 남겨뒀고 아래 코멘트는 전반적인 리뷰이니 한 번 읽어보세요!

---

1. API 요청에 대한 처리(Ex. error, loading 등)가 조금 더 명확하게 있으면 좋을거 같아요!
지금은 백엔드와 협업을 하는게 아니다보니 어쩔 수 없는 부분이 있지만 앞으로 개발하시면서 비슷한 처리를 밥 먹듯이 하게 될테니 이 기회에 연습해보시는걸 추천드립니다!

2. 올바른 변수 이름 짓기는 개발자에게 가장 중요하고 어려운 작업 중 하나입니다.
변수명을 지을 때 조금 더 고민해서 다른 사람이 변수명만으로 어떤 역할을 하는 데이터인지 한 번에 파악할 수 있으면 지어보는걸 추천드립니다!

jinhok96 committed:
chore: createBrowserRouter 적용
	modified:   src/App.js

jinhok96 committed:
chore: value -> addLinkValue로 변경
	modified:   src/components/Header/AddLinkBar.js

jinhok96 committed:
refactor: import 경로를 절대 경로로 변경
	modified:   .eslintrc.js
	new file:   jsconfig.json
	modified:   package-lock.json
	modified:   package.json
	modified:   src/App.js
	modified:   src/components/Common/AddFolderButton.js
	modified:   src/components/Common/AddLInkButton.js
	modified:   src/components/Common/Button.js
	modified:   src/components/Common/ErrorMessage.js
	modified:   src/components/Common/FlaotingAddFolderButton.js
	modified:   src/components/Common/LoginButton.js
	modified:   src/components/Common/Option.js
	modified:   src/components/Common/SortingButton.js
	modified:   src/components/Footer/Footer.js
	modified:   src/components/Header/AddLinkBar.js
	modified:   src/components/Header/FolderHeaderContent.js
	modified:   src/components/Header/Gnb.js
	modified:   src/components/Header/Header.js
	modified:   src/components/Header/SharedHeaderContent.js
	modified:   src/components/Main/Card.js
	modified:   src/components/Main/CardList.js
	modified:   src/components/Main/Main.js
	modified:   src/components/Main/SearchBar.js
	modified:   src/components/Main/SortingSection.js
	modified:   src/index.js
	modified:   src/pages/FolderPage/FolderPage.js
	modified:   src/pages/HomePage/HomePage.js
	modified:   src/pages/SharedPage/SharedPage.js
	modified:   src/services/api.js

jinhok96 committed:
chore: result 객체 내부에 error가 없을 경우 수정
	modified:   src/hooks/useFetch.js

jinhok96 committed:
chore: api url 환경 변수로 분리
	new file:   .env
	modified:   .eslintrc.js
	modified:   jsconfig.json
	modified:   src/hooks/useFetch.js
	modified:   src/index.js

jinhok96 committed:
refactor: css 경로 절대 경로로 변경
	modified:   src/index.css

jinhok96 committed:
chore: 이름 규칙에 맞게 변경
	modified:   src/components/Header/Gnb.js

jinhok96 committed:
chore: || -> ?? 로 변경
	modified:   src/components/Header/Gnb.js

jinhok96 committed:
chore: API_URL을 함수 내부로 이동
	modified:   src/hooks/useFetch.js

jinhok96 committed:
chore: useLocation destructuring 적용
	modified:   src/components/Header/Gnb.js

jinhok96 committed:
refactor: css 경로 절대 경로로 수정
	modified:   src/App.js
	modified:   src/assets/styles/utility.css
	modified:   src/components/Common/AddFolderButton.js
	modified:   src/components/Common/AddLInkButton.js
	modified:   src/components/Common/Button.js
	modified:   src/components/Common/ErrorMessage.js
	modified:   src/components/Common/FlaotingAddFolderButton.js
	modified:   src/components/Common/LoginButton.js
	modified:   src/components/Common/Option.js
	modified:   src/components/Common/SortingButton.js
	modified:   src/components/Footer/Footer.js
	modified:   src/components/Header/AddLinkBar.js
	modified:   src/components/Header/FolderHeaderContent.js
	modified:   src/components/Header/Gnb.js
	modified:   src/components/Header/Gnb.module.css
	modified:   src/components/Header/Header.js
	modified:   src/components/Header/SharedHeaderContent.js
	modified:   src/components/Main/Card.js
	modified:   src/components/Main/CardList.js
	modified:   src/components/Main/Main.js
	modified:   src/components/Main/SearchBar.js
	modified:   src/components/Main/SortingSection.js
	modified:   src/index.css
	modified:   src/index.js
	modified:   src/pages/FolderPage/FolderPage.js
	modified:   src/pages/HomePage/HomePage.js
	modified:   src/pages/SharedPage/SharedPage.js

jinhok96 committed:
refactor: 스타일링 구조 수정하여 고장난 레이아웃 복구
	modified:   src/assets/styles/utility.css
	modified:   src/components/Common/FlaotingAddFolderButton.js
	modified:   src/components/Footer/Footer.js
	modified:   src/components/Footer/Footer.module.css
	modified:   src/components/Header/Gnb.js
	modified:   src/components/Header/Gnb.module.css
	modified:   src/pages/FolderPage/FolderPage.js
	modified:   src/pages/FolderPage/FolderPage.module.css

jinhok96 committed:
refactor: 고장난 레이아웃 수정
	modified:   src/components/Common/AddFolderButton.js
	modified:   src/components/Header/AddLinkBar.js
	modified:   src/components/Header/AddLinkBar.module.css
	modified:   src/components/Main/SortingSection.js
	modified:   src/components/Main/SortingSection.module.css

jinhok96 committed:
chore: OptionList 마진 수정
	modified:   src/components/Common/Option.module.css
	modified:   src/components/Main/SortingSection.js
	modified:   src/components/Main/SortingSection.module.css

jinhok96 committed:
chore: footer를 화면 맨 밑에 고정
	modified:   src/assets/styles/utility.css
	modified:   src/pages/FolderPage/FolderPage.js
	modified:   src/pages/SharedPage/SharedPage.js

jinhok96 committed:
chore: 케밥 버튼 컴포넌트 생성
	new file:   src/assets/images/kebab.svg
	new file:   src/assets/images/star.svg
	new file:   src/components/Common/KebabButton.js
	new file:   src/components/Common/KebabButton.module.css

jinhok96 committed:
feat: 케밥 버튼 추가
refactor: flex-row, flex-col관련 속성 분리
	modified:   src/assets/styles/colors.css
	modified:   src/assets/styles/utility.css
	modified:   src/components/Common/KebabButton.js
	modified:   src/components/Common/Option.js
	modified:   src/components/Footer/Footer.js
	modified:   src/components/Header/AddLinkBar.js
	modified:   src/components/Header/Gnb.js
	modified:   src/components/Main/Card.js
	modified:   src/components/Main/Card.module.css
	modified:   src/components/Main/SortingSection.js
	modified:   src/components/Main/SortingSection.module.css

jinhok96 committed:
chore: Main 여백 스타일 수정
	modified:   src/components/Main/Main.js

jinhok96 committed:
chore: StarButton 컴포넌트 생성
	modified:   src/components/Common/KebabButton.js
	new file:   src/components/Common/StarButton.js
	new file:   src/components/Common/StarButton.module.css

jinhok96 committed:
feat: 스타 버튼 추가
	modified:   src/components/Main/Card.js
	modified:   src/components/Main/Card.module.css

jinhok96 committed:
chore: api 호출 관련 error, loading 로직 개선
	modified:   src/hooks/useFetch.js
	modified:   src/services/api.js

jinhok96 committed:
chore: sample api 분리
	modified:   src/components/Header/SharedHeaderContent.js
	modified:   src/services/api.js

jinhok96 committed:
feat: SortingSection에 api 적용
	modified:   src/assets/styles/utility.css
	modified:   src/components/Common/AddFolderButton.js
	modified:   src/components/Header/Gnb.js
	modified:   src/components/Header/SharedHeaderContent.js
	modified:   src/components/Header/SharedHeaderContent.module.css
	modified:   src/components/Main/CardList.js
	modified:   src/components/Main/SortingSection.js
	modified:   src/hooks/useFetch.js
	modified:   src/services/api.js

jinhok96 committed:
chore: button -> folder로 변경
	modified:   src/components/Main/SortingSection.js

jinhok96 committed:
chore: CardList에 api 적용
	modified:   src/components/Main/CardList.js
	modified:   src/components/Main/CardList.module.css
	modified:   src/components/Main/SortingSection.js
	modified:   src/pages/FolderPage/FolderPage.js

jinhok96 committed:
chore: 콘솔 출력 제거
	modified:   src/components/Main/CardList.js

jinhok96 committed:
feat: 링크가 없을 경우 메시지 출력
	modified:   src/components/Main/CardList.js
	modified:   src/components/Main/SortingSection.js
	modified:   src/pages/FolderPage/FolderPage.js

jinhok96 committed:
chore: id type을 number로 통일
	modified:   src/components/Main/CardList.js
	modified:   src/components/Main/SortingSection.js
	modified:   src/pages/FolderPage/FolderPage.js

jinhok96 committed:
fix: mobile 크기에서 AddFolderButton이 사라지지 않는 문제 수정
	modified:   src/components/Common/AddFolderButton.js

SeolJaeHyeok commented on .env:
```diff
@@ -0,0 +1 @@
+REACT_APP_API_URL = https://bootcamp-api.codeit.kr/api/
```
👍👍

.env 파일은 민감 정보를 담고 있기 때문에 github 같은 원격 저장소에는 올리지 않는게 좋아요!
따라서 .gitignore 파일에 추가 해주시고 원격 저장소에는 지워주세요~

SeolJaeHyeok commented on .eslintrc.js:
```diff
@@ -0,0 +1,121 @@
+          },
+          {
+            pattern: 'pages/**/*',
+            group: 'internal',
+            position: 'after',
+          },
+          {
+            pattern: 'services/**/*',
+            group: 'internal',
+            position: 'after',
+          },
+          {
+            pattern: 'utils/**/*',
+            group: 'internal',
+            position: 'after',
+          },
+        ],
+        'newlines-between': 'always',
+        alphabetize: {
+          order: 'asc',
+        },
+      },
+    ],
+  },
+  settings: {
+    'import/resolver': {
+      alias: {
+        map: [
+          ['assets', './src/assets'],
+          ['components', './src/components'],
+          ['context', './src/context'],
+          ['hooks', './src/hooks'],
+          ['pages', './src/pages'],
+          ['services', './src/services'],
+          ['utils', './src/utils'],
+        ],
+        extensions: ['.js', '.jsx', '.json'],
+      },
+    },
+  },
```
플러그인까지 👍👍

SeolJaeHyeok commented on jsconfig.json:
```diff
@@ -0,0 +1,14 @@
+{
+  "compilerOptions": {
+    "baseUrl": "src",
+    "paths": {
+      "assets/*": ["assets/*"],
+      "components/*": ["components/*"],
+      "context/*": ["context/*"],
+      "hooks/*": ["hooks/*"],
+      "pages/*": ["pages/*"],
+      "services/*": ["services/*"],
+      "utils/*": ["utils/*"]
+    }
+  }
+}
```
절대 경로 설정 좋습니다!

SeolJaeHyeok commented on package.json:
```diff
@@ -19,7 +19,7 @@
     "test": "react-scripts test",
     "eject": "react-scripts eject",
-    "lint": "eslint ."
+    "lint": "eslint . --ext .js,.jsx,.ts,.tsx"
```
타입스크립트까지 미리 대응하신 의욕 훌륭합니다

SeolJaeHyeok commented on src/App.js:
```diff
@@ -1,15 +1,19 @@
-import { BrowserRouter as Router, Route, Routes } from 'react-router-dom';
-import Shared from './pages/Shared/Shared';
+
+const router = createBrowserRouter([
+  { path: '/', element: <HomePage /> },
+  { path: 'shared', element: <SharedPage /> },
+  { path: 'folder', element: <FolderPage /> },
+]);
```
👍
react-router-dom이 v6로 넘어오면서 정말 다양한 기능을 지원하게 됐는데, createBrowserRouter도 그 중 하나에요.
지금처럼 선언적으로 경로를 다룰 수 있을 뿐 아니라, 페이지 진입 전 데이터 페칭이 가능하다던가 nested도 쉽게 설정 가능하니 적극적으로 사용해보시는걸 추천드려요!

SeolJaeHyeok commented on src/components/Header/Gnb.js:
```diff
@@ -1,63 +1,76 @@
   // 로고 클릭 시 페이지 최상단으로 이동
   const handleLogoClick = () => {
     window.scrollTo(0, 0);
   };
```
`로고 클릭 시 페이지 최상단으로 이동`이라는 기능에서 `페이지 최상단으로 이동`은 커스텀 훅이나 유틸로 분리가 가능할 거 같아요!

그렇게 되면 `로고 클릭 시`라는 특정 상황에서만 사용하는게 아니라 여러가지 상황에서 더 범용적으로 사용이 가능하겠죠?!?

SeolJaeHyeok commented on src/hooks/useFetch.js:
```diff
@@ -0,0 +1,42 @@
+function useFetch(apiUrl) {
+  const API_URL = process.env.REACT_APP_API_URL;
+
+  // data, error 초기값은 비어있음, loading 초기값은 로딩 중인 상태
+  const [data, setData] = useState(null);
+  const [loading, setLoading] = useState(true);
+  const [error, setError] = useState(null);
+
+  // apiUrl 또는 API_URL이 변경될 경우 실행
+  useEffect(() => {
+    const fetchData = async () => {
+      try {
+        const url = new URL(apiUrl, API_URL);
+        const response = await fetch(url);
+        const result = await response.json();
+
+        const responseError = result?.error ?? 'Fetch response failed';
+
+        if (!response.ok) {
+          throw new Error(`User response error: ${responseError}`);
+        }
+
+        setData(result);
+      } catch (errorData) {
+        const errorString = errorData.toString();
+        setError(errorString);
+      } finally {
+        setLoading(false); // 로딩 끝
+      }
+    };
+
+    // 함수 호출
+    fetchData();
+  }, [apiUrl, API_URL]);
+
+  // 객체 반환
+  return { data, loading, error };
+}
+
+export default useFetch;
```
👍👍
공통화 좋습니다! 어디서든 get요청이면 이 Hook을 유용하게 사용할 수 있을거 같아요!

SeolJaeHyeok commented on src/services/api.js:
```diff
@@ -1,131 +1,27 @@
-  const response = await fetch(url);
-  const result = await response.json();
-  const responseError = result.error;
+export function getFoldersApiUrl() {
+  const url = 'users/1/folders';
+  return url;
+}
```
함수로 분리하는 시도는 좋았습니다!

함수로 분리하는 목적은 가독성 증가, 로직 분리, 각 함수의 역할 최소화로 사이드 이펙트를 줄이기, 공통화 등 여러가지 이유가 있을거 같아요
말씀드리고 싶은 부분은 `함수를 분리한다`는 행위의 목적을 고민해보시면 좋을거 같아요.
모든 방식은 trade off가 존재하기 때문에 `모든걸 함수로 분리하겠어!`와 같이 함수를 집착적으로 분리하기보다는 장점과 단점을 생각하고 함수를 분리하는 연습을 해보시는걸 추천드려요!

SeolJaeHyeok reviewed (commented):
역시나 잘 해주셨네요 ㅎㅎ

추가적으로 코멘트 남겼으니 확인해주세요~

jinhok96 committed:
chore: .env .gitignore에 추가
	modified:   .gitignore

jinhok96 committed:
chore: .env 삭제
	deleted:    .env

jinhok96 committed:
chore: 필요없는 주석 제거
	modified:   src/components/Header/Gnb.js

jinhok96 committed:
chore: scrollToTop 유틸 함수 추가
	new file:   src/utils/scrollToTop.js

jinhok96 committed:
chore: 페이지 상단으로 이동하는 기능을 scrollToTop로 분리
	modified:   src/components/Header/Gnb.js

jinhok96 committed:
refactor: api 주소를 함수에서 상수로 변경
- 유동적인 주소를 가진 api는 함수형 유지, 통일성 위해 이름은 상수형
	modified:   src/components/Header/Gnb.js
	modified:   src/components/Header/SharedHeaderContent.js
	modified:   src/components/Main/CardList.js
	modified:   src/components/Main/SortingSection.js
	modified:   src/services/api.js

SeolJaeHyeok reviewed (commented):
고생 많으셨습니다! 이번 PR은 머지하도록 할게요~!
